File: day2/go/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

var availableCubes = map[string]int{
	"red":   12,
	"green": 13,
	"blue":  14,
}

func splitOffGameId(line string) (int, string) {
	noPrefixLine := strings.TrimPrefix(line, "Game ")

	groups := strings.Split(noPrefixLine, ":")
	gameNumStr, restOfLine := groups[0], strings.Join(groups[1:], "")

	gameNum, err := strconv.Atoi(gameNumStr)
	if err != nil {
		log.Fatalf("cannot convert %s to int", gameNumStr)
	}

	return gameNum, restOfLine
}

func getGameSets(line string) []string {
	sets := strings.Split(line, ";")

	for i := 0; i < len(sets); i++ {
		sets[i] = strings.TrimSpace(sets[i])
	}

	return sets
}

func getSetCubes(set string) map[string]int {
	setCubes := make(map[string]int)

	numColorGroups := strings.Split(set, ", ")

	for _, numColor := range numColorGroups {
		elems := strings.Split(numColor, " ")

		quantStr, color := elems[0], elems[1]

		quant, err := strconv.Atoi(quantStr)
		if err != nil {
			log.Fatalf("cannot convert %s to int", quantStr)
		}

		setCubes[color] = quant
	}

	return setCubes
}

func validCubeQuantities(setCubes map[string]int) bool {
	for color, quant := range setCubes {
		if availableCubes[color] < quant {
			return false
		}
	}

	return true
}

func getArgs() (int, string) {
	if len(os.Args) != 3 {
		log.Fatal("Usage: go run main.go <1 or 2> <full or example>")
	}

	part, err := strconv.Atoi(os.Args[1])
	if err != nil || (part != 1 && part != 2) {
		log.Fatal("First argument must be either 1 or 2")
	}

	mode := os.Args[2]
	if mode != "example" && mode != "full" {
		log.Fatal("Second argument must be either example or full")
	}

	return part, mode
}

func readInput(name string) *os.File {
	path := fmt.Sprintf("../input/%s", name)
	file, err := os.Open(path)
	if err != nil {
		log.Fatalf("failed to open file: %s", err)
	}

	return file
}

func main() {
	part, mode := getArgs()

	file := readInput(mode)
	defer file.Close()

	if part == 1 {
		sumValidIds := 0

		scanner := bufio.NewScanner(file)
		for scanner.Scan() {
			line := scanner.Text()

			gameId, restOfLine := splitOffGameId(line)

			gameSets := getGameSets(restOfLine)
			gameIsValid := true
			for _, gameSet := range gameSets {
				setCubes := getSetCubes(gameSet)

				if !validCubeQuantities(setCubes) {
					gameIsValid = false
					break
				}
			}

			if gameIsValid {
				sumValidIds += gameId
			}
		}

		fmt.Println(sumValidIds)
	}

	if part == 2 {
		sumCubePowers := 0

		scanner := bufio.NewScanner(file)
		for scanner.Scan() {
			line := scanner.Text()

			_, restOfLine := splitOffGameId(line)

			gameSets := getGameSets(restOfLine)
			maxQuantCubes := make(map[string]int)
			for _, gameSet := range gameSets {
				setCubes := getSetCubes(gameSet)

				for color, quant := range setCubes {
					if maxQuantCubes[color] < quant {
						maxQuantCubes[color] = quant
					}
				}

			}

			powerOfCubes := 1
			for _, quant := range maxQuantCubes {
				powerOfCubes *= quant
			}

			sumCubePowers += powerOfCubes
		}

		fmt.Println(sumCubePowers)
	}
}

File: day2/python/main.py
import math
import sys
from collections import defaultdict
from typing import Iterator

AVAILABLE_CUBES = {
    "red": 12,
    "green": 13,
    "blue": 14,
}


def split_off_game_id(line: str) -> tuple[int, str]:
    game_num, game_info = line.removeprefix("Game ").split(":", 1)

    return int(game_num), game_info


def get_game_sets(line: str) -> list[str]:
    sets = [s.strip() for s in line.split(";")]

    return sets


def get_set_cubes(a_set: str) -> dict[str, int]:
    # You may find this too readable
    # set_cubes = {}

    # num_color_groups = a_set.split(", ")

    # for num_color in num_color_groups:
    #     elems = num_color.split(" ")

    #     quant, color = int(elems[0]), elems[1]

    #     set_cubes[color] = quant

    # return set_cubes

    # In that case, use this
    return {
        color: int(quant)
        for num_color in a_set.split(", ")
        for quant, color in (num_color.split(" "),)
    }


def is_valid_cube_quantities(set_cubes: dict[str, int]) -> bool:
    return all(AVAILABLE_CUBES[color] >= quant for color, quant in set_cubes.items())


def is_valid_game(game_sets: list[str]) -> bool:
    # Sane approach
    # for game_set in game_sets:
    #     set_cubes = get_set_cubes(game_set)

    #     if not is_valid_cube_quantities(set_cubes):
    #         return False

    # return True

    # "Short" approach
    return all(
        is_valid_cube_quantities(get_set_cubes(game_set)) for game_set in game_sets
    )


def get_args() -> tuple[int, str]:
    if len(sys.argv) != 3:
        print("Usage: python main.py <1 or 2> <example or full>")
        sys.exit(1)

    part = int(sys.argv[1])
    if part not in (1, 2):
        print("First argument must be either 1 or 2")
        sys.exit(1)

    mode = sys.argv[2]
    if mode not in ("example", "full"):
        print("Second argument must be either example or full")
        sys.exit(1)

    return part, mode


def read_input(name: str) -> Iterator[str]:
    with open(f"../input/{name}", "r") as f:
        for line in f:
            yield line.rstrip()


def main():
    part, mode = get_args()

    if part == 1:
        sum_valid_ids = 0

        for line in read_input(mode):
            game_id, game_info = split_off_game_id(line)

            game_sets = get_game_sets(game_info)

            if is_valid_game(game_sets):
                sum_valid_ids += game_id

        print(sum_valid_ids)

    if part == 2:
        sum_cube_powers = 0

        for line in read_input(mode):
            _, game_info = split_off_game_id(line)
            game_sets = get_game_sets(game_info)

            max_quant_cubes = defaultdict(int)
            for game_set in game_sets:
                set_cubes = get_set_cubes(game_set)

                for color, quant in set_cubes.items():
                    max_quant_cubes[color] = max(max_quant_cubes[color], quant)

            power_of_cubes = math.prod(max_quant_cubes.values())
            sum_cube_powers += power_of_cubes

        print(sum_cube_powers)


if __name__ == "__main__":
    main()
